### PR TITLE
QRadar better retry, fix duplicates and polling

### PR DESCRIPTION
**Pull request by ilaner**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://jira-hq.paloaltonetworks.local/browse/XSUP-15649?filter=-1
https://jira-hq.paloaltonetworks.local/browse/XSUP-13239
https://jira-hq.paloaltonetworks.local/browse/XSUP-15397?filter=-1
https://jira-hq.paloaltonetworks.local/browse/XSUP-15244?filter=-1

## Description
- Improved retry mechanism handling of failed events fetch.
- Improved missing events messages in the QRadar widgets.
- Fixed an error message when **create-search** command failed.
- Fixed duplicate incidents creation.
- Try to poll events immediately after creating a search in **qradar-search-retrieve-events** command.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3403]:
```suggestion
    now_millis: int = int(datetime.now().timestamp() * 1000)
```
or `now_milliseconds`

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3411]:
```suggestion
            f'START {offense_start_time} STOP {now_millis}'
```

**Review by tkatzir — APPROVED** [review]
Looks good.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (DeanArbel) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1658]: Are there no `events` or were there none fetched?

Comment (ilaner) [reply Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1658]: none fetched, fixed message

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:20]: Shouldn't this be 30 seconds, to round up if needed?

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1660]:
```suggestion
        print_debug_msg(f'No events were fetched for offense {offense_id}. '
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1540]: From what I could tell, `create_events_search` doesn't raise an exception,
so you'll only iterate once?

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1538]: This:
`if num_of_failures == max_retries:`
(and its block) should be out of the loop.

**Review by tkatzir — COMMENTED** [review]
I believe I see some potential issues with the loop.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (ilaner) [reply Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:20]: 20 seconds should be enough (the third request should succeed in all cases)

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1548]:
```suggestion
    print_debug_msg(f'Reached max retries for creating a search for offense: {offense_id}. Returning error.')
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1655]:
```suggestion
        print_debug_msg(f'No events were fetched for offense {offense_id}. Retrying in {FAILURE_SLEEP} seconds.')
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1659]: What if mirroring isn't enabled?
Should we advise the user to enable mirroring?

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1882]:
```suggestion
            f'Successfully Created {len(incidents)} incidents. Incidents created: {[incident["name"] for incident in incidents]}')
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:2416]:
```suggestion
            raise DemistoException(f'Could not create events search for offense_id: {offense_id}')
```
Distinguish the exception and log messages.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3534]:
```suggestion
                          outputs=search_command_results.outputs if search_command_results else None,
```

**Review by tkatzir — COMMENTED** [review]
Looks good!
Waiting with approval for when it's ready.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py
Comment (ilaner) [reply Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1659]: If mirroring isn't enabled, the events will not be fetched at all. We already advise enabling mirroring in this case

**Review by tkatzir — APPROVED** [review]
Looks good to me, os I've approved,
But please wait for @DeanArbel's approval as well, before merging.

**Review by DeanArbel — APPROVED** [review]
Good job, see my 2 comments

File: Packs/QRadar/ReleaseNotes/2_2_10.md
Comment (DeanArbel) [inline Packs/QRadar/ReleaseNotes/2_2_10.md:13]:
```suggestion
- Improved retry handling when not all events are fetched.
- QRadar widgets now display better messages when not all events are fetched.
- Fixed an error message when **qradar-search-create** command failed.
- Fixed an issue where duplicate incidents were created.
- The command **qradar-search-retrieve-events** will try to poll events immediately after creating a search.

#### Scripts
##### QRadarMirroringEventsStatus
- Improved messages when not all events are fetched.
- Updated the Docker image to: *demisto/python3:3.10.5.31928*.
```

File: Packs/QRadar/ReleaseNotes/2_2_10.md
Comment (DeanArbel) [inline Packs/QRadar/ReleaseNotes/2_2_10.md:7]:
```suggestion
- Improved the error message when **qradar-search-create** command fails.
```

**Pull request merged**
